### PR TITLE
feat(java): extract public Fury methods to BaseFury

### PR DESCRIPTION
**Pull request by Munoon**
Extract public Fury method to the `BaseFury` interface and implement them in `ThreadPoolFury` and `ThreadLocalFury`, so end users can avoid using `execute()` and other utility methods.

Off topic: further steps might be changing `BaseFury` interface name to `Fury` while existing class renaming to something like `FuryImpl`. `FuryBuilder` should return the new `Fury` interface, so we will hide system methods like `xwriteRef` and others. WDYT?

### CODE REVIEW

File: java/fury-core/src/main/java/org/apache/fury/BaseFury.java [hunk 1]
```diff
@@ -64,9 +71,16 @@ public interface BaseFury {
    */
   <T> void registerSerializer(Class<T> type, Class<? extends Serializer> serializerClass);
 
+  void registerSerializer(Class<?> type, Serializer<?> serializer);
```
Comment (chaokunyang) [inline java/fury-core/src/main/java/org/apache/fury/BaseFury.java:74]: The serializer is bind to a Fury instance, we may not be able to create a serializer object for registration unless we change the serializer interfere. Could we change this to register(Class type, Class serializer class)?

File: java/fury-core/src/main/java/org/apache/fury/BaseFury.java [hunk 2]
```diff
@@ -55,6 +59,9 @@ public interface BaseFury {
    */
   void register(Class<?> cls, Short id, boolean createSerializer);
 
+  /** register class with given type tag which will be used for cross-language serialization. */
+  void register(Class<?> cls, String typeTag);
```
Comment (chaokunyang) [inline java/fury-core/src/main/java/org/apache/fury/BaseFury.java:63]: This API is used for clang serialization, we removed it in our new xlang serialization spec in #1413 . Could we remove this API? We can add  new API when new protocol are implemented

File: java/fury-core/src/main/java/org/apache/fury/BaseFury.java (hunk 1 above)
Comment (Munoon) [reply java/fury-core/src/main/java/org/apache/fury/BaseFury.java:71]: There might be a singleton serializer. I've personally create some of them in my project. In this case they will be applied to each existing and each new Fury instances in `ThreadLocalFury` and `ThreadPoolFury`.
Maybe this is something that worth mentioning in the documentation.

However, if you didn't like it, we can remove this method and make user register singleton serializers using `execute` method. In this case, there is more chances that user will know what he is doing.

Comment (chaokunyang) [reply java/fury-core/src/main/java/org/apache/fury/BaseFury.java:71]: I see, thanks. This is reasonable to add to BaseFury

**Review by chaokunyang — APPROVED** [review]
LGTM

**Pull request merged**